Update README.md

zapster opened the pull request:
Basically a copy&paste from https://wiki.openjdk.java.net/display/Graal/Instructions. Fixes #1. Alternative to #78.

dougxc commented:
I'd like to make the migration of documentation from OpenJDK to README.md be somewhat atomic. Let's chat tomorrow to co-ordindate this.

eregon commented:
:+1:

dougxc commented:
The plan is to migrate [Debugging](https://wiki.openjdk.java.net/display/Graal/Debugging), [Eclipse](https://wiki.openjdk.java.net/display/Graal/Eclipse) and [Publications](https://wiki.openjdk.java.net/display/Graal/Publications+and+Presentations) from the [OpenJDK wiki](https://wiki.openjdk.java.net/display/Graal) to a `docs` directory in this repo as `Debugging.md`, `Eclipse.md` and `Publications.md` respectively. The remaining content on the OpenJDK wiki will either be moved to the [Truffle repo](https://github.com/graalvm/graal-core) or deleted.

zapster commented:
Added Debugging, Eclipse and Publications files. Note that some of them might be outdated. I can fix issues if there are specific suggestions but maybe we should do this in a separate PR.
Also, I've removed links to Graal JavaDoc since they are 404ing.

zapster commented:
FYI: I stopped the travis build since its only documentation and I have to update to HEAD anyway.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
 ## Building Graal
```
We should start with a brief description of what Graal is before launching into the build instructions. Can you please add this:

Graal is a dynamic compiler written in Java that integrates with the HotSpot JVM. It has a focus on high performance and extensibility. In addition, it provides optimized performance for [Truffle](https://github.com/graalvm/mx) based languages running on the JVM.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+    export PATH=$PWD/mx:$PATH
+
+The Graal code depends on JVMCI ([JVM Compiler Interface](https://bugs.openjdk.java.net/browse/JDK-8062493)) and the Truffle API, both of which need to be cloned along with Graal. To do this, create a working directory (e.g. named graal) into which both Graal and JVMCI will be cloned and then use mx to clone both:
+
```
Please use markdown formatting for code sections.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+    cd graal-core
+    mx
+
+The mx command ensures the JVMCI version in sync with Graal is cloned. Changing to the graal-core directory which informs mx that the focus of development (called the primary suite) is Graal. All subsequent mx commands should be executed from this directory.
```
"which informs" -> "informs"

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
-
-These configurations aim to match as closely as possible the
-VM(s) included in the OpenJDK binaries one can download.
+To simplify Graal development, a separate Python tool called mx has been co-developed. This tool must be downloaded and put onto your PATH:
```
Make mx be a link to mx repo.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+
+    mx spull
+
+More detailed information on mx is available [here](https://github.com/graalvm/mx).
```
Remove this sentence.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+    mx build
+    mx vm
+
+The first time it's run, the mx build step above will present you with a dialogue asking which VM you will be using by default. The choices are described in below. If you are going to be working with Truffle, you should select the server option, otherwise select the jvmci option.
```
Inline mx commands should use markdown formatting (e.g., `mx build`).

rschatz replied:
I think for command sequences this format should be used:
```
mkdir graal
cd graal
git clone https://github.com/graalvm/graal-core.git
cd graal-core
mx
```

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+### VM types
+
+ * The server configuration is a standard HotSpot VM that includes the runtime support for Graal but uses the existing compilers for normal compilation (e.g., when the interpreter threshold is hit for a method). Compilation with Graal is only done by explicit requests to the Graal API. This is how Truffle uses Graal.
+ * The jvmci configuration is a VM where normal compilations are performed by Graal. This VM will bootstrap Graal itself at startup unless the -XX:-BootstrapJVMCI VM option is given. Note that if tiered compilation is enabled, Graal will be used at the last tier while C1 will be used for the first compiled tiers.
```
Remove the inaccurate statement about bootstrapping.

dougxc commented on README.md:
```diff
@@ -1,80 +1,60 @@
+
+This will generate both Eclipse and NetBeans project configurations. Further information on how to import these project configurations into Eclipse can be found [here](docs/Eclipse.md).
+
+The Graal code base includes the [Ideal Graph Visualizer](http://ssw.jku.at/General/Staff/TW/igv.html) which is very useful in terms of visualizing Graal's intermediate representation (IR). You can get a quick insight into this tool by running the commands below. The first command launches the tool and the second runs one of the unit tests included in the Graal code base with extra options to mak ...
```
IGV is of course only available if JVMCI is not a binary import. We will have to address this soon but it's ok to pretend it's part of graal-core for now.

dougxc commented on docs/Debugging.md:
```diff
@@ -0,0 +1,130 @@
+
+## Debugging
```
We don't need the Debugging title here - it's redundant.

dougxc commented on docs/Debugging.md:
```diff
@@ -0,0 +1,130 @@
+
+All the parts of Graal written in Java can be debugged with a standard Java debugger. While debugging with Eclipse is explicitly supported, it should not be too hard to adapt these instructions for another debugger.
+
+The mx eclipseinit command not only creates Eclipse project configurations but also creates an Eclipse launch configuration (in mx/eclipse-launches/attach-localhost-8000.launch) that can be used to debug all Graal code running in the VM. This launch configuration requires you to start the VM with the -d global option which puts the VM into a state waiting for a remote debugger to attach to port 8 ...
```
Use Markdown formatting: `mx eclipseinit`

Please check for all other missing uses of MD formatting.

dougxc commented on docs/Debugging.md:
```diff
@@ -0,0 +1,130 @@
+
+### Logging
+
+In addition to IDE debugging, Graal includes support for *printf* style debugging. The simplest is to place temporary usages of System.out where ever you need them. Assuming that you are developing in Eclipse and have the Checkstyle plugin installed, the source lines containing `System.out` will be highlighted in red but the code will still work. The Graal gate will prevent such code from being i ...
```
Remove the statement about Checkstyle warning and the gate checking it.

dougxc commented on docs/Debugging.md:
```diff
@@ -0,0 +1,130 @@
+
+In addition to IDE debugging, Graal includes support for *printf* style debugging. The simplest is to place temporary usages of System.out where ever you need them. Assuming that you are developing in Eclipse and have the Checkstyle plugin installed, the source lines containing `System.out` will be highlighted in red but the code will still work. The Graal gate will prevent such code from being i ...
+
+For more permanent logging statements, use the `Debug.log()` method that is part of debug scope mechanism. A (nested) debug scope is entered using the scope method Debug. For example:
```
"using the `Debug.scope(...)` method."

dougxc commented on docs/Debugging.md:
````diff
@@ -0,0 +1,130 @@
+}
+```
+
+The Debug.log statement will send output to the console if "CodeInstall" is matched by the `-G:Log` GraalVM option. The matching logic for this option is described in DebugFilter. As mentioned in the javadoc, the same matching logic also applies to the `-G:Dump`, `-G:Time` and `-G:Meter` options.
````
add link to DebugFilter javadoc source.

dougxc commented:
@thomaswue can you please review this change so Josef can merge it.
@zapster please squash the commits before merging.

thomaswue commented on README.md:
```diff
@@ -1,80 +1,59 @@
+cd graal
+git clone https://github.com/graalvm/graal-core.git
+cd graal-core
+mx
```
Shouldn't this be "mx sclone".

dougxc replied:
No, the suite references are resolved the first time mx is executed and I'd prefer to omit mention of s* commands unless unavoidable.

zapster commented:
@dougxc yes, will do once we have a final version
